_posts/2015-12-14-simply-secure-rails-api.markdown

```diff
@@ -0,0 +1,499 @@
+block cipher algorithm mentioned above requires its' input to be an EXACT multiple of the block size. Setting this to 1 allows for variance in the size of the data being encrypted. 
+If you're very interested: read the full [documentation](http://ruby-doc.org/stdlib-1.9.3/libdoc/openssl/rdoc/OpenSSL/Cipher.html#method-i-padding-3D)
+
+---
+
+#### Line 6
+```ruby
+encrypted =  cipher.update(self.uuid) + cipher.final
+```
+This takes the actual data you want to encrypt, in this case the user's 
+UUID, and turns is into something that's actually ready to be deciphered. For practical purposes, calling .final a procedural event in the cipher's life-cycle. 
+[Documentation](http://docs.ruby-lang.org/en/trunk/OpenSSL/Cipher.html#method-i-final)
+
+---
+
+#### Line 7
+```ruby
+Base64.urlsafe_encode64(encrypted).encode('utf-8')
+```
+This is the actual data we want to transform the finalized cipher into so that it can be passed safely via a query string or form data. Only over SSL, of course :)
+
+---
+
+After our authentication application's user model has an encrypted method we can modify the: 
+
+##### Authentication Application's User's controller
+
+```ruby
+module Api
+  module V1
+    class LoginController < ApiController
+      
+      def create # POST: /api/v1/login
+        @user = User.find_by(email: params[:email])
+
+        if @user.present? && @user.valid_password?(params[:password])
+          respond_with: @user.as_json(methods: [:encrypted])
+        else 
+          head 404
+        end
+      end
+
+    end
+  end
+end
+```
+
+Now that our authentication application is prepped to return encrypted UUIDs (when provided a valid email and password), we
+can create our Social Networking application here's some boilerplate:
+
+##### Our Social Network's user class
+
+```ruby
+class User < ActiveRecord::Base
+  # Schema
+  # Name: string
+  # UUID: UUID (primary key)
+  # Email
+
+  def self.create_from_authentication_service(email, password)
+    begin
+      connection =  Faraday.new(url: "https://id.wework.io") do |faraday|
+                      faraday.adapter Faraday.default_adapter    
+                      faraday.headers['Content-Type'] = 'application/json'
+                    end
+
+      response = connection.post("/api/v1/login/", {
+                    email: email,
+                    password: password
+                  }.to_json)
+
+      parsed_body = JSON.parse(response.body)
+
+      User.create({
+        uuid:  parsed_body["user"]["uuid"],
+        email: parsed_body["user"]["email"],
+        name:  parsed_body["user"]["name"],
+      })
+    rescue 
+      false
+    end
+  end
+end
+```
+
+##### Our Social Network login controller
+
+```ruby
+class LoginController < ApplicationController
+  
+  def new # Get /login
+  end
+
+  def create  # Post/login
+
+    user = User.create_from_authentication_service(params[:email], params[:password])
+
+    if user
+      cookies.signed[:uuid] = user.uuid
```

Comment:
We should use a token, not the uuid (it can't be revoked).
Also, do we actually need this code to make the point? I'd rather show the least we need when it comes to security